packages/engine/src/networking/serialization/DataWriter.ts

```diff
@@ -81,10 +85,115 @@ export const writeVector4 = (vector4: Vector4SoA) => (v: ViewCursor, entity: Ent
   return (changeMask > 0 && writeChangeMask(changeMask)) || rewind()
 }
 
+// Writes a compressed Quaternion value to the network stream. This function uses the "smallest three"
+// method, which is well summarized here: http://gafferongames.com/networked-physics/snapshot-compression/
+export const writeCompressedRotation = (vector4: Vector4SoA) => (v: ViewCursor, entity: Entity) => {
+  const rewind = rewindViewCursor(v)
+  const writeChangeMask = spaceUint8(v)
+  const rewindUptoChageMask = rewindViewCursor(v)
+  let changeMask = 0
+  let b = 0
+
+  // Todo: writeVector4 here? how to handle rewind than.
+  // Write to store
+  changeMask |= writePropIfChanged(v, vector4.x, entity) ? 1 << b++ : b++ && 0
+  changeMask |= writePropIfChanged(v, vector4.y, entity) ? 1 << b++ : b++ && 0
+  changeMask |= writePropIfChanged(v, vector4.z, entity) ? 1 << b++ : b++ && 0
+  changeMask |= writePropIfChanged(v, vector4.w, entity) ? 1 << b++ : b++ && 0
+
+  // Write to ViewCursor
+  if (changeMask > 0) {
+    rewindUptoChageMask()
+    let maxIndex = 0
+    let maxValue = Number.MIN_VALUE
+    let sign = 1
+
+    // Determine the index of the largest (absolute value) element in the Quaternion.
+    // We will transmit only the three smallest elements, and reconstruct the largest
+    // element during decoding.
+    for (let i = 0; i < 4; i++) {
+      let element = getVector4IndexBasedComponentValue(vector4, entity, i) as number
+      let abs = Math.abs(element)
+      if (abs > maxValue) {
+        // We don't need to explicitly transmit the sign bit of the omitted element because you
+        // can make the omitted element always positive by negating the entire quaternion if
+        // the omitted element is negative (in quaternion space (x,y,z,w) and (-x,-y,-z,-w)
+        // represent the same rotation.), but we need to keep track of the sign for use below.
+        sign = element < 0 ? -1 : 1
+
+        // Keep track of the index of the largest element
+        maxIndex = i
+        maxValue = abs
+      }
+    }
+
+    let a = 0
+    let b = 0
+    let c = 0
+
+    if (maxIndex === 0) {
+      a = vector4.y[entity]
+      b = vector4.z[entity]
+      c = vector4.w[entity]
+    } else if (maxIndex === 1) {
+      a = vector4.x[entity]
+      b = vector4.z[entity]
+      c = vector4.w[entity]
+    } else if (maxIndex === 2) {
+      a = vector4.x[entity]
+      b = vector4.y[entity]
+      c = vector4.w[entity]
+    } else {
+      a = vector4.x[entity]
+      b = vector4.y[entity]
+      c = vector4.z[entity]
+    }
+
+    // Multiply with QUAT_MAX_RANGE & FLOAT_PRECISION_MULT before compression so that values are
+    // capped to required range([-0.707107,+0.707107]) and precision(three decimal places)
+    a *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+    b *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+    c *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+
+    maxIndex = maxIndex | 0
+
+    const compress = (value: number) => {
```

Comment:
This function is also be created anew every time writeCompressedRotation is called